### PR TITLE
feat: Add summarizing and quoting capabilities to 05-assistive-chatbot

### PR DESCRIPTION
**Pull request by yoomlam**
## Ticket

Partly resolves https://navalabs.atlassian.net/browse/DST-218

## Changes

Populate prototype code in `05-assistive-chatbot` with the summarizing and quoting capabilities from `02-household-queries` chatbot.

Update handling `settings` for LLM clients and did some general refactoring.

## Context for reviewers

This provides baseline summarizing and quoting capabilities for the prototype.

## Testing

Testing instructions and expected behavior:
1. `pip install -r requirements.txt`
1. Update `.env`
2. To clear and populate the DB, run: `rm -rf chroma_db && ./ingest-guru-cards.py`
1. Try it on the command line, for example:
```
CHATBOT_LOG_LEVEL=INFO CHAT_ENGINE=Summaries    LLM_MODEL_NAME='openai :: gpt-3.5-turbo-instruct' LLM_MODEL_NAME_2='openai :: gpt-3.5-turbo-instruct' RETRIEVE_K=2 ./cmdline.py

CHATBOT_LOG_LEVEL=INFO CHAT_ENGINE=Summaries-DSPy LLM_MODEL_NAME='dspy :: gpt-3.5-turbo-instruct' LLM_MODEL_NAME_2='openai :: gpt-3.5-turbo-instruct' RETRIEVE_K=2 ./cmdline.py
```
1. Try it via the web app (To ensure summaries are generated correctly, pick `*instruct` model for both LLMs):
```
./chatbot-chainlit.py
```
   - Open a browser to http://localhost:8000/
   - Open the `Settings panel` by clicking on the icon to the left of the bottom text input field.
   - Choose "Summaries" as the `Chat Mode` (synonymous with `CHAT_ENGINE` for the `cmdline.py` app) and pick `openai :: gpt-3.5-turbo-instruct` as the `Primary LLM` and `LLM Model for summarizer`. Adjust the two `Temperature...` settings to play with the responses.
   - click `Confirm`
   - type in any question and see derived questions, Guru cards, their summaries, and quotes:
image

### CODE REVIEW

File: 05-assistive-chatbot/ingest-guru-cards.py [hunk 1]
```diff
@@ -1,87 +1,35 @@
...
-from langchain_community.vectorstores import Chroma
 
 import chatbot
-from chatbot import guru_cards, utils
+from chatbot import guru_cards, vector_db
 from chatbot.ingest.text_splitter import TextSplitter
 
 logger = logging.getLogger(f"chatbot.{__name__}")
 
-
```
Comment (yoomlam) [inline 05-assistive-chatbot/ingest-guru-cards.py:18]: moved the following to `vector_db.py`

File: 05-assistive-chatbot/chatbot/engines/v2_household_engine.py [hunk 2]
```diff
@@ -1,56 +1,194 @@
...
+        for card in dq_entry.retrieved_cards:
+            if card not in card_to_dqs:
+                card_to_dqs[card] = set()
+            card_to_dqs[card].add(derived_question)
+
+        for card, quote in zip(dq_entry.retrieved_cards, dq_entry.retrieved_chunks):
+            if card not in card_to_quotes:
+                card_to_quotes[card] = set()
+            if quote != card:
+                card_to_quotes[card].add(quote)
```
Comment (yoomlam) [inline 05-assistive-chatbot/chatbot/engines/v2_household_engine.py:154]: If the quote == card title, then don't use it as a quote b/c the card title will already be shown in the UI and the title/question is not particular helpful as a quote.

File: 05-assistive-chatbot/chatbot/vector_db.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        logger.info("Creating Vector DB: %s", self.folder)
+        return Chroma(
+            embedding_function=self.embeddings_model,
+            persist_directory=self.folder,
+            # Must use collection_name="langchain" -- https://github.com/langchain-ai/langchain/issues/10864#issuecomment-1730303411
+            collection_name="langchain",
+        )
+
+
+ingest_vectordb_wrapper = LocalLangchainChromaVectorDb()
```
Comment (yoomlam) [inline 05-assistive-chatbot/chatbot/vector_db.py:62]: `ingest_vectordb_wrapper` provides the same default settings for use by `ingest-guru-cards.py` and the chatbot.

File: 05-assistive-chatbot/chatbot-chainlit.py [hunk 4]
```diff
@@ -101,18 +124,44 @@ async def message_submitted(message: cl.Message):
...
+                "",
+            ]
+            indented_quotes = [q.strip().replace("\n", "\n   ") for q in card.quotes]
+            cards_resp += [f"\n   Quote:\n   ```\n   {q}\n   ```" for q in indented_quotes]
+            cards_resp += ["</details>", ""]
+
+    return {
+        "content": "\n".join(resp + dq_resp + cards_resp),
+        "elements": [
+            # cl.Text(name="Derived Questions", content="\n".join(dq_resp), display="side"),
```
Comment (ccheng26) [inline 05-assistive-chatbot/chatbot-chainlit.py:161]: Is this still needed?

File: 05-assistive-chatbot/chatbot/engines/v2_household_engine.py [hunk 5]
```diff
@@ -1,56 +1,194 @@
...
-    def gen_response(self, query):
-        response = self.client.generate_reponse(query)
-        # TODO: decompose query
-        # TODO: create summaries
-        return response
+        # Use the same vector DB configuration as ingest-guru-cards.py
+        self.vectordb_wrapper = vector_db.ingest_vectordb_wrapper
+        self.retrieve_k = int(settings.pop("retrieve_k"))
+
+        # TODO: ingestigate if this should be set to true
```
Comment (ccheng26) [inline 05-assistive-chatbot/chatbot/engines/v2_household_engine.py:66]: investigate*

**Review by ccheng26 — APPROVED** [review]
no blocking issues, lgtm!

**Pull request merged**